### ISSUE TITLE
i

### ISSUE THREAD

**Issue by goosui** [mbebenita/Broadway#162]

**Comment by goosui** [comment]
I tried it again with play.js

Still unable to decode P frames

//demo code
===============================================================
<!DOCTYPE html>
<html>
<head>
	<meta name="viewport" content="width=320, initial-scale=1"/>
	<title>h264 streaming</title>
	<style type="text/css">
		body {
			background: #333;
			text-align: center;
			margin-top: 10%;
		}
		#videoCanvas {
			/* Always stretch the canvas to 640x480, regardless of its
			internal size. */
			width: 640px;
			height: 480px;
		}
	</style>
</head>
<body>
	
	
	<script type="text/javascript" src="Decoder.js"></script>
  <script type="text/javascript" src="YUVCanvas.js"></script>
  <script type="text/javascript" src="Player.js"></script>
	
	<script type="text/javascript">
				
		var player = new Player({size: {
        width: 640,
        height: 320
         }});
    
		document.body.appendChild(player.canvas);
 
 
	var xhr = new XMLHttpRequest(); 
	function GetImage(){
		var timestamp = (new Date()).valueOf();
		xhr.open('GET', 'http://192.168.1.152/enc0.h264?t='+timestamp, true); 
		xhr.send();
	}
	xhr.type="HEAD";	
	xhr.responseType = 'arraybuffer'; 
	xhr.onload = function(e) {  
		if (this.status == 200) {
			var uInt8 = new Uint8Array(this.response);		
			if(uInt8[0]==0){
				console.log('---------丢掉 uInt8[0]='+uInt8[0]);	
				return;
			}
			console.log('--------1-uInt8[0]='+uInt8[0]);
			if(uInt8[0]==85){
				uInt8[0] = 0;	

					var  naltype;
				  if (uInt8[4] == 0x65) {
					naltype = "I frame";	
				  }
				  else if (uInt8[4] == 0x61) {				
					
				  }
				  else if (uInt8[4] == 0x67) {
					naltype = "SPS";				
					
				  }
				  else if (uInt8[4] == 0x68) {
					naltype = "PPS";
					
				  }	
				player.decode(uInt8);
			}
		}
	}; 


	setInterval("GetImage()",50);

		
	</script>
</body>
</html>

**Comment by annazhelt** [comment]
I encode raw video frames one by one using libavcodec on the backend. Then, I send the encoded frame to frontend over a websocket. All I have in the frontend is a VideoManager class which contains the following code:
```
constructor(){
    this.player = new broadway.Player({
      webgl: true,
      useWorker: true,
      workerFile: 'static/js/Decoder.js'
    });
}

updateFrame(frameData){
  if(frameData.length === 0){
    return;
  }
  this.player.decode(new Uint8Array(frameData)); 
}
``` 
I don't even need to split the frames into NAL units. I open a new websocket connection and attach an onMessage handler to the connection:
```
onMessage: (evt) => {
  this.video.updateFrame(evt.data);
}
```
I have also noticed in your code you assume P frames are 0x61. It should be 0x41 which is 61 in decimal.

**Comment by goosui** [comment]
@annazhelt
 Yes,the transmission of my example is the same as yours. I'm going to study your explanation on the top.

**Comment by goosui** [comment]
@annazhelt

`
constructor(){
    this.player = new broadway.Player({
      webgl: true,
      useWorker: true,
      workerFile: 'static/js/Decoder.js'
    });
}
`

“broadway“ Where do you get it? Do you define yourself? Sorry I am not find broadway.js  from the author's source code.

**Comment by annazhelt** [comment]
I installed the broadway-player node module through npm. Then, I added the following import statement to my code: `import broadway from 'broadway-player';` This gives you access to Player, Decoder, and Canvas.

**Comment by goosui** [comment]
can you help me to test this code .I am use this can only decode some frame.

can you replace your websocket url to test?

thank you.

==================================================================
[websocketDemo.zip](https://github.com/mbebenita/Broadway/files/1409818/websocketDemo.zip)



<!doctype html>
<html>
<head>
  <script type="text/javascript" src="Decoder.js"></script>
  <script type="text/javascript" src="YUVCanvas.js"></script>
  <script type="text/javascript" src="Player.js"></script>
</head>
<body>
<canvas id="canvasId" width="640" height="480"></canvas>
<div id="rgbId">
</div>

<button id="closeSendId">
<script>
	//第一种方式
	var canvas = document.getElementById('canvasId');
	canvas.width = 1280;
	canvas.height =720;
	canvas.style.zIndex = 8;
	canvas.style.position = "absolute";	
	var ctx2D = canvas.getContext('2d');

	var img = document.createElement("img");
    img.onload = function(e) {
		ctx2D.drawImage(this,0,0,1280,720);
		console.log("---------draw");
        window.URL.revokeObjectURL(this.src); 
    };
	
	var pDecoder = new Decoder({
	  rgb:true,
	});
	pDecoder.onPictureDecoded =  function(buffer, width, height, infos){
		var imageData = ctx2D.getImageData(0, 0, width, height);
		imageData.data.set(buffer);
        ctx2D.putImageData(imageData, 0, 0);
	};
	
	
	var url = "ws://192.168.1.100:80/main.ws?user=root";//your ip	
	if ('WebSocket' in window) {
		console.log("create Web socket url="+url);
		ws = new WebSocket(url);
	} else if ('MozWebSocket' in window) {
		console.log("create MozWeb socket");
		ws = new MozWebSocket(url);
	} else {
		ws=new WebSocket("ws://localhost:80");
		console.log('Unsupported');
	}
	
	ws.onopen = function() {
		console.log("开启websocket关于main.ws");	
	};		
	ws.binaryType="arrayBuffer";
	ws.onmessage = function(event) {
		console.log('Receive message: ' + event.data);
		var reader = new FileReader();
		reader.readAsArrayBuffer(event.data);
		reader.onload = function (e) {
			console.info(reader.result); //ArrayBuffer {}
			var buf = new Uint8Array(reader.result);
			console.log('length=' + buf.length);
			console.log('buf[0]=' + buf[0]);
			console.log('buf[1]=' + buf[1]);
			console.log('buf[2]=' + buf[2]);
			console.log('buf[3]=' + buf[3]);
			
			var  naltype;
			  if (buf[4] == 0x65) {
				naltype = "I frame";	
			  }
			  else if (buf[4] == 0x61) {
				naltype = "P frame";
			  }
			  else if (buf[4] == 0x67) {
				naltype = "SPS";
			  }
			  else if (buf[4] == 0x68) {
				naltype = "PPS";
			  }	
			  console.log("------naltype="+naltype);
			
			  pDecoder.decode(buf);
		}
	}
	
	function getBinDate(){
		 var copyU8 = new Uint8Array(32); 
		 ws.send(copyU8);
	}
	setTimeout("getBinDate()",70);

	

</script>
</body>
</html>

**Comment by soliton4** [comment]
only using decoder.js should work. i am sorry i do not have the time to debug your code for you.
if you have a specific question feel free to ask it

**Comment by scaret** [comment]
Try:
`uInt8[3]=0;`
=>
`uInt8[3]=1;`

**Comment by chenchenwuai** [comment]
@goosui   老铁，你的这个问题解决了吗？
